src/driver/mod.rs

```diff
@@ -0,0 +1,293 @@
+use bytes::BytesMut;
+use std::time::Duration;
+use tokio::io::{AsyncReadExt, AsyncWriteExt};
+
+use crate::parser::{Response, Value};
+use crate::{parser, MemcacheError, PoolConnection};
+
+const EMPTY_SPACE_BYTES: &[u8] = b" ";
+const NEW_LINE_BYTES: &[u8] = b"\r\n";
+const NO_REPLY_BYTES: &[u8] = b" noreply\r\n";
+const COMMAND_DELETE: &[u8] = b"delete ";
+const COMMAND_TOUCH: &[u8] = b"touch ";
+const COMMAND_VERSION: &[u8] = b"version\r\n";
+
+/// Storage command
+#[derive(Debug)]
+pub enum StorageCommand {
+    /// "set" means "store this data".
+    Set,
+    /// "add" means "store this data, but only if the server *doesn't* already
+    /// hold data for this key".
+    Add,
+    /// "replace" means "store this data, but only if the server *does*
+    /// already hold data for this key".
+    Replace,
+}
+
+impl From<StorageCommand> for &'static [u8] {
+    fn from(c: StorageCommand) -> &'static [u8] {
+        match c {
+            StorageCommand::Set => b"set ",
+            StorageCommand::Add => b"add ",
+            StorageCommand::Replace => b"replace ",
+        }
+    }
+}
+
+/// <command name> <key> <flags> <exptime> <bytes> [noreply]\r\n
+///
+///
+/// - "STORED\r\n", to indicate success.
+///
+/// - "NOT_STORED\r\n" to indicate the data was not stored, but not
+/// because of an error. This normally means that the
+/// condition for an "add" or a "replace" command wasn't met.
+///
+/// - "EXISTS\r\n" to indicate that the item you are trying to store with
+/// a "cas" command has been modified since you last fetched it.
+///
+/// - "NOT_FOUND\r\n" to indicate that the item you are trying to store
+/// with a "cas" command did not exist.
+pub async fn storage<K>(
+    mut conn: PoolConnection<'_>,
+    command: StorageCommand,
+    key: K,
+    flags: u32,
+    expiration: impl Into<Option<Duration>>,
```

Comment:
nit, I'd probably avoid mixing generic types and anonymous generic types, hence replace `impl Into<Option<Duration>>,` with an `E` type and add `E: Into<Option<Duration>>,` to where clause.